text/0002-poet2-consensus.md

```diff
@@ -0,0 +1,566 @@
+          digest of (OPKhash | PPK).
+	* Verify the nonce in the AVR is equal to `WaitCertId_{n}`, namely the
+          digest of the most recently committed block. It may be that the sender
+          has not seen `WaitCertId_{n}` yet and could be sending
+          `WaitCertId_{n'}` where `n'<n`.
+          In this case the sender should be urged to updated his/her
+          view of the ledger by appending the new blocks and retry. It could
+          also happen that the receiving validator has not seen
+          `WaitCertId_{n}` in which case he/she should try to update his/her
+          view of the ledger and verify again.
+	* Verify MRENCLAVE value within quote is equal to PoET\_MRENCLAVE (there
+      could be more than one allowed value).
+	* Verify basename in the quote is equal to distributed ledger Basename.
+	* Verify attributes field in the quote has the allowed value (normally
+          the enclave must be in initialized state and not be a debug enclave).
+ 
+   If the join request is not self attested (Self Attestation is NOT enabled in
+   IAS API): (OPK, PPK, quote)
+
+   * Create AEP with quote:
+
+      * isvEnclaveQuote: base64 encoded quote
+
+ 3. Send AEP to IAS. The IAS sends back a signed AVR.
+ 4. Verify received AVR attests to validity of both quote and
+    save EPID Pseudonym.
+ 5. Verify ``report_data`` field within the quote contains the SHA256 digest of
+	(OPKhash | PPK).
+ 6. Verify MRENCLAVE value within quote is equal to PoET\_MRENCLAVE (there could
+	be more than one allowed value).
+ 7. Verify basename in the quote is equal to distributed ledger Basename.
+ 8. Verify attributes field in the quote has the allowed value (normally the
+    enclave must be in initialized state and not be a debug enclave).
+   
+   If the verification fails, exit.
+
+   If the verification succeeds but the SGX platform identified by the EPID
+   Pseudonym in the quote has already signed up, ignore the join request, exit.
+
+   If the verification succeeds:
+
+   * Pass sign-up certificate of new participant (OPK, EPID Pseudonym, PPK,
+     current `WaitCertId_{n}` to upper layers for registration in
+     EndPoint registry.
+   * Goto 1
+     
+# Leader Election
+### Wall Clock and Chain Clock
+The PoET 2.0 algorithm maintains two clocks, the _Wall Clock_ and the
+_Chain Clock_.
+
+The Wall Clock (`WC`) measures the time elapsed since the arrival of the
+'Sync Block', typically the first block on the chain. It is maintained by the
+validator on the 'untrusted' portion of the PoET code. The Chain Clock (`CC`)
+measures the cumulative `WaitTime` of all the blocks in the chain (each
+fork will have its own `CC`).
+
+Upon receiving the Sync Block, `WC` and `CC` are initialized to 0. 
+
+>Note 1: In practice, the WC may be calculated by recording the system time at 
+>the moment of the arrival of the Sync Block and subsequently subtracting this 
+>timestamp from the current time.
+

```

Comment:
Its probably worth receiving several blocks and computing an average over those blocks to ensure that you don't favor a low-latency neighbor or a cheating neighbor.